system/jlib/jutil.cpp

```diff
@@ -3349,10 +3367,13 @@ const char * generatePassword(StringBuffer &pwd, int pwdLen)
     const char alphaUC[] = "ABCDEFGHIJKLMNOPQRSTUVWXYZ";
     const char alphaLC[] = "abcdefghijklmnopqrstuvwxyz";
     const char numeric[] = "0123456789";
-    const char symbol[] = "~@#$%^*()_-+={[}]:,.?";
+    const char symbol[] = "~@#%^*_-+{[}]:,.?";
+
+    unsigned randomNumber;
+    generateRandom((const void *)&randomNumber, sizeof(randomNumber));
+    srand(randomNumber);
```

Comment:
I think this probably needs a bit more discussion (possibly including the code in lnuid).  A couple of items to discuss:

* windows needs to be supported.
* should /dev/urandom only be used for seeding a random number generator (e.g., create an instance of std::mt19937 and seed it with values read from /dev/urandom. (see quote below:)
* rand isn't thread safe

"While some safety margin above that minimum is reasonable, as a guard against flaws in the CPRNG algorithm, no cryptographic primitive available today can hope to promise more than 256 bits of security, so if any program reads more than 256 bits (32 bytes) from the kernel random pool per invocation, or per reasonable reseed interval (not less than one minute), that should be taken as a sign that its cryptography is not skillfully implemented"

Reply:
There are two mechanisms "/dev/random" and "/dev/urandom".

urandom actually does close to what you suggest already.  It uses the entropy to seed a "cryptographically secure" pseudorandom number generator (CSPRNG).  It then reseeds at a certain rate.  I think the default is 60 sec.  I think the quote above is talking about limits in trusting certain key sizes to be truly random.  But I've seen arguments declaring that urandom is pretty secure.

Using mt19937 would mostly be switching out the repeated use of the internal urandom CSPRNG.  Delivering similar results although perhaps it is a better algorithm?  The mt19937 algorithm should only repeat every (219937 – 1) times.  Would need to investigate if it were really better than whats used in urandom.

But anyway, in this case using mt19937 (and as Russ is trying to do with rand()) as a seed should be fine, but I don't think it fixes what the quote is talking about above, because it just replaces the CPRNG with another generator but I could be wrong if the algorithm is better.

Whether we use urandom or not I guess Russ should not be using rand if it's not thread safe.  So either using urandom to seed mt19937 or just using urandom should be good enough?

As for lnguid, they (Charles, David and Co.) did some pretty extensive analysis trying to determine how frequently the same id would be produced by their algorithm using urandom, but it wouldn't hurt to think it over again.

Reply:
Bottom line, replace rand since it's not threadsafe.

Replacement could either be just urandom, or mt19937 seeded with urandom.

Reply:
https://en.wikipedia.org/wiki/Cryptographically-secure_pseudorandom_number_generator#:~:text=Every-,CSPRNG,-should%20satisfy%20the